Journal aggregation (SQL based approach)

NobodysNightmare opened the pull request:
## OpenProject work package

https://community.openproject.org/work_packages/20686
## Dependencies

This PR depends on https://github.com/opf/openproject/pull/3247
## Description

This PR adds the ability to obtain aggregated journals. Journals are aggregated if they succeed eachother on the same journalizable and were issued within a defined timespan.

This PR chooses a different implementation approach than https://github.com/opf/openproject/pull/3216, which I do not want to throw away (yet).
## TODO
- [x] lookup a journal for editing (by id, exposed via `notes_id`)
- ~~determine earliest/latest `created_at`~~
- [x] create index over (`journalized_type`, `journalized_id`, `version`)
  - extracted into https://github.com/opf/openproject/pull/3247
- [x] use the actual setting to determine the timespan
- [x] final performance test after rebasing onto https://github.com/opf/openproject/pull/3247

NobodysNightmare committed:
remove dead code

NobodysNightmare committed:
first spike for aggregatable journals

NobodysNightmare committed:
make aggregation capable for MySQL and PSQL

- also give the time threshold a useful meaning (number of seconds)

NobodysNightmare committed:
add basic spec for aggregated journals

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-sql-journal-aggregation.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-sql-journal-aggregation.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare committed:
test non-aggregation after time gap

NobodysNightmare committed:
change aggregated_journals on AaJ to be a method

- a "real" relation involves cascading updates that are not intended for _aggregated_ journals

NobodysNightmare committed:
styling

NobodysNightmare committed:
remove obsoleted helper

NobodysNightmare committed:
add specs for not aggregating comments

NobodysNightmare committed:
Enhance AggregatedJournals to correctly merge journals with notes

Aggregation should now follow the rules previously defined for merging journals (see code comments).
Jens (@ulferts) and I developed the SQL for this. I hope the explanatory comments make it
a quite readable SQL experience.

myabc commented on lib/plugins/acts_as_journalized/lib/acts_as_journalized.rb:
```diff
@@ -96,44 +89,11 @@ def acts_as_journalized(options = {}, &block)
             :updated_on << :updated_at << :lock_version << :lft << :rgt
 
-          prepare_journaled_options(journal_hash)
+          journal_hash = prepare_journaled_options(journal_hash)
```
@NobodysNightmare changes here won't merge cleanly into `feature/rails4` branch.

NobodysNightmare replied:
Lets see what happens :)

I could leave out most of the changes here, since the greatest part is only refactoring:
- finding out that great parts are not actually being used and can therefore be deleted
- changing a method to not modify its input (this line)

If you think this is a bigger problem I can also not do those changes...

myabc commented on lib/plugins/acts_as_journalized/lib/acts_as_journalized.rb:
```diff
@@ -172,6 +132,12 @@ def journalized_event_hash(options)
       end
+
+      # We can't use a :has_many relation as this would try to delete journals upon deleting the
+      # journalized object (there is nothing like "dependent: ignore").
```
@NobodysNightmare `dependent: :nullify`?

NobodysNightmare replied:
Nullify will also change the journals (setting the foreign key reference to `NULL`). I have some problems here:
1. The `has_many :journals`  would have deleted those records anyway (aggregated journals are only a subset of those)
2. The way we are aggregating the journals might even break when trying to delete the journals (it did for me once, but that was before @ulferts and I refactored the SQL, not sure what would happen now)
3. We even blend some information from different journals, so that makes it an even worse idea to delete records based on that

I now feel that `has_many`, might for multiple reasons not be the right choice here...

NobodysNightmare committed:
add more specs for aggregated journals

NobodysNightmare committed:
add PostgreSQL-compatible group numbering

NobodysNightmare committed:
make SQL compliant for both MySQL and PSQL

NobodysNightmare committed:
expose id of "master" as notes_id

- make spec assertions more detailed by comparing all the attributes

NobodysNightmare committed:
use the actual setting for the aggregation time

NobodysNightmare committed:
override :initial? for AggregatedJournal

the simple approach of looking at the version number does not work, because aggregated journals can have gaps in their versions (and there might be no version 1)

NobodysNightmare committed:
better PSQL compatibility for aggregated journals

Azure7111 commented on app/models/journal/aggregated_journal.rb:
```diff
@@ -0,0 +1,168 @@
+# Similar to regular Journals, but under the following circumstances journals are aggregated:
+#  * they are in temporal proximity
+#  * they belong to the same resource
+#  * they were performed by the same user
```
Suggestion: created instead of performed

Azure7111 commented on app/models/journal/aggregated_journal.rb:
```diff
@@ -0,0 +1,168 @@
+# "open" a new aggregation, since we do not want to merge comments in any way.
+# The term "aggregation" means the following when applied to our journaling:
+#  * ignore/hide old journal rows (since every journal row contains a full copy of the journaled)
+#    object, dropping intermediate rows will just increase the diff of the following journal
```
I might just be confused here: How does dropping intermediate rows increase the diff?

If I get this point correctly then the fact being stated is that in a group of journals which can be aggregated only the properties of the newest journal will be used. Is that correct?

NobodysNightmare replied:
At least my use of parenthesis was broken, I repaired that (so this diff will soon be outdated).

I think you've got it: We use the latest journal and compare it to its predecessor to get a diff. By removing intermediate rows, more time between the journal and its predeccessor passed. Thereby the diff gets larger*

*terms and conditions apply

Azure7111 commented on app/models/journal/aggregated_journal.rb:
```diff
@@ -0,0 +1,168 @@
+    def default_scope
+      # Using the roughly aggregated groups from :sql_rough_group we need to merge journals
+      # where an entry with empty notes follows an entry containing notes, so that the notes
+      # from the main entry are taken, while the remaining information from the more recent entry
```
Suggestion: "..., while the remaining information is taken from the more recent entry."

Azure7111 commented on app/models/journal/aggregated_journal.rb:
```diff
@@ -0,0 +1,168 @@
+
+    # MySQL requires some initialization to be performed before being able to count the groups.
+    # This method allows to inject further FROM sources to achieve that in a single SQL statement.
+    # Sadly MySQL requires the whole statement to be wrapped in parenthesis, while PostreSQL
```
Totally hyped to see that new "PostreSQL" everybody tells me about ;)

Azure7111 commented on app/models/journal/aggregated_journal.rb:
```diff
@@ -0,0 +1,168 @@
+    end
+
+    # The "group_number" required in :sql_rough_group has to be generated differently depending on
+    # the DBMS used. This method returns the apropriate statement to be used inside a select to
```
Typo: appropriate
Also: What do you think about capitalizing "SELECT"?

Azure7111 commented on app/models/journal/aggregated_journal.rb:
```diff
@@ -0,0 +1,168 @@
+    # Returns a SQL condition that will determine whether two entries are too far apart (temporal)
+    # to be considered for aggregation. This takes the current instance settings for temporal
+    # proximity into account.
+    def sql_beyond_aggregation_time?(predecessor, successor)
```
This method name should technically probably not end with ? since you are not returning a boolean _value_ but rather a boolean _condition_. Though I am totally fine with how it is.

NobodysNightmare replied:
I had the same thought, but also came to the same conclusion ^^

NobodysNightmare committed:
clarify explanations

typos, wording and stuff

Azure7111 commented on app/models/journal/aggregated_journal.rb:
```diff
@@ -0,0 +1,91 @@
+  self.table_name = 'journals'
+
+  class << self
+    def default_scope
```
Is there some AR magic going on here which dictates the name of this method? Because otherwise I'd think that its not very descriptive.

NobodysNightmare replied:
Yep...

That means when you do `AggregatedJournal.all` (or `AggregatedJournal.where(...)`, etc) the scope here will automatically apply.

NobodysNightmare committed:
allow to fetch aggregated journals for a journable

- this should be faster (or at least easier to optimize) than applying a WHERE to the whole scope
- also starting to diverge from using classic ActiveRecord in aggregated journals

Azure7111 commented on spec/models/journal/aggregated_journal_spec.rb:
```diff
@@ -0,0 +1,202 @@
+
+        context 'adding a second comment' do
+          before do
+            expect(work_package.update_by!(new_author, notes: notes)).to be_truthy
```
Using the same comment here threw me off a little. Suggestion: Use a `notes` text which smartly explains why it leads to the journals not being merged.

Or: Give the test a longer name, similar to https://github.com/opf/openproject/pull/3230/files#diff-d8deafe0017a1f18ef44bc573ea59375R150

Could this help to understand the situation a little better if the following test were to fail? (Think about what might be written as the reason for test failure)

Azure7111 commented on spec/models/journal/aggregated_journal_spec.rb:
```diff
@@ -0,0 +1,202 @@
+    before do
+      work_package.status = FactoryGirl.build(:status)
+      work_package.save!
+      work_package.journals.second.created_at += 1.day # one day delay should always be long enough
```
Why don't you use something like `Setting.journal_aggregation_time_minutes + 1`?

Azure7111 commented:
:+1:  LGTM so far. Let me know when you think its ready.

NobodysNightmare committed:
remove the AR portion of aggregated journals

- this prohibits doing stupid things like saving aggregated journals
- should make clear that aggregated journals are not real journals...

NobodysNightmare committed:
make specs more readable

NobodysNightmare commented:
@Azure7111 see my latest changes.

I feel ready if travis agrees.

Azure7111 commented on app/models/journal/aggregated_journal.rb:
```diff
@@ -0,0 +1,222 @@
+    # To be able to self-join results of this statement, we add an additional column called
+    # "group_number" to the result. This allows to compare a group resulting from this query with
+    # its predecessor and successor.
+    def sql_rough_group(journable, uid)
```
You might want to have some tests covering the `journable` parameter here.

NobodysNightmare committed:
add spec for single journable aggregation

Azure7111 commented:
Will merge when green :+1:

Azure7111 commented:
Have a look at that postgres failure, it looks legit.

https://travis-ci.org/opf/openproject/jobs/72416757

NobodysNightmare commented:
No it does not...

@ulferts says this is a well known flickering test
